src/libraries/System.Linq/src/System/Linq/CountBy.cs

```diff
@@ -0,0 +1,60 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+
+namespace System.Linq
+{
+    public static partial class Enumerable
+    {
+        public static IEnumerable<KeyValuePair<TKey, int>> CountBy<TSource, TKey>(this IEnumerable<TSource> source, Func<TSource, TKey> keySelector, IEqualityComparer<TKey>? keyComparer = null) where TKey : notnull
+        {
+            if (source is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.source);
+            }
+            if (keySelector is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.keySelector);
+            }
+
+            return Core(source, keySelector, keyComparer);
+
+            static IEnumerable<KeyValuePair<TKey, int>> Core(IEnumerable<TSource> source, Func<TSource, TKey> keySelector, IEqualityComparer<TKey>? keyComparer)
```

Comment:
Following naming convention elsewhere this should be called `CountByIterator`, see for example:

https://github.com/dotnet/runtime/blob/ba10d55c99968ab577a9d96f19fbb2dbddbc83a4/src/libraries/System.Linq/src/System/Linq/Zip.cs#L87-L111

I would also recommend not making it a local method so that any exception stacktraces show up cleaner.

Reply:
Method renamed and local methods removed.